compiler/rustc_mir_transform/src/dest_prop.rs

```diff
@@ -208,22 +208,23 @@ impl<'tcx> MirPass<'tcx> for DestinationPropagation {
             // This is the set of merges we will apply this round. It is a subset of the candidates.
             let mut merges = FxHashMap::default();
 
-            for (src, candidates) in candidates.c.iter() {
-                if merged_locals.contains(*src) {
+            for (src, candidates) in candidates.c.drain() {
+                if merged_locals.contains(src) {
                     continue;
                 }
                 let Some(dest) =
-                    candidates.iter().find(|dest| !merged_locals.contains(**dest)) else {
+                    candidates.into_iter().find(|(dest, _)| !merged_locals.contains(*dest)) else {
                         continue;
                 };
                 if !tcx.consider_optimizing(|| {
                     format!("{} round {}", tcx.def_path_str(def_id), round_count)
                 }) {
                     break;
                 }
-                merges.insert(*src, *dest);
-                merged_locals.insert(*src);
-                merged_locals.insert(*dest);
+                merged_locals.insert(src);
+                merged_locals.insert(dest.0);
+                merges.insert(src, dest.clone());
+                merges.insert(dest.0, dest);
```

Comment:
Do I understand correctly that the first entry `merges.insert(src, dest.clone())` will result only in local replacements (`src` to `dest.0`), and the second `merges.insert(dest.0, dest)` only in write removals to local `dest.0`?

Maybe those two operations should be described by separate data structures?